components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -6,82 +6,419 @@
+      DVLOG(2) << __func__ << ": NTP SR test data will be loaded"
+               << " from local path at: "
+               << forced_local_path.LossyDisplayName();
+      OnComponentReady(false, forced_local_path);
+    } else {
+      CheckSuperReferralComponent();
+    }
+  }
+#endif
+}
 
+void NTPBackgroundImagesService::RegisterSponsoredImagesComponent() {
   const std::string locale =
       brave_ads::LocaleHelper::GetInstance()->GetLocale();
-  if (const auto& data = GetRegionalComponentData(
-          brave_ads::LocaleHelper::GetCountryCode(locale))) {
-    RegisterNTPBackgroundImagesComponent(cus, data.value(),
-        base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
-                            weak_factory_.GetWeakPtr()));
+  const auto& data = GetSponsoredImagesComponentData(
+      brave_ads::LocaleHelper::GetCountryCode(locale));
+  if (!data) {
+    DVLOG(2) << __func__ << ": Not support NTP SI component for " << locale;
+    return;
+  }
+
+  DVLOG(2) << __func__ << ": Start NTP SI component";
+  RegisterNTPBackgroundImagesComponent(
+      component_update_service_,
+      data->component_base64_public_key,
+      data->component_id,
+      base::StringPrintf("NTP Sponsored Images (%s)", data->region.c_str()),
+      base::BindRepeating(&NTPBackgroundImagesService::OnComponentReady,
+                          weak_factory_.GetWeakPtr(),
+                          false));
+}
+
+void NTPBackgroundImagesService::CheckSuperReferralComponent() {
+  if (base::FeatureList::IsEnabled(features::kBraveNTPBrandedWallpaperDemo)) {
+    RegisterDemoSuperReferralComponent();
+    return;
+  }
+
+  const auto* value = local_pref_->Get(
+      prefs::kNewTabPageCachedSuperReferralComponentInfo);
+  if (IsValidSuperReferralComponentInfo(*value)) {
+    RegisterSuperReferralComponent();
+    return;
+  }
+
+  if (local_pref_->FindPreference(
+          prefs::kNewTabPageCachedSuperReferralComponentInfo)->
+              IsDefaultValue()) {
+    // For existing users, we could set it as non SR because SR is only applied
+    // for fresh user.
+    // At first fresh launch, we could determine after this is SR or not after
+    // downloading mapping table. but browser could be shutdown between getting
+    // promocode and mapping table. If this happens, this install will be act
+    // as a non SR install forever. To resolve that situation,
+    // |kNewTabPageCheckingMappingTableInProgress| introduced.
+    // Even if |kReferralCheckedForPromoCodeFile| is true, we will try to check
+    // mapping table again if needed.
+    // |kNewTabPageCheckingMappingTableInProgress| is set to true when mapping
+    // table is requested and set to false when browser gets it.
+    // If both |kReferralCheckedForPromoCodeFile| and
+    // |kNewTabPageCheckingMappingTableInProgress|, browser had some trouble
+    // while getting mapping table at first fresh launch.
+    if (local_pref_->GetBoolean(kReferralCheckedForPromoCodeFile) &&
```

Comment:
NOTE:  If https://github.com/brave/brave-core/pull/5144 is merged before this PR,  consider newly added `kReferralInitialization` also here.